### PR TITLE
[EDA] Prevent Deletion of Test Record  if There are Associated Test Score Records

### PR DESCRIPTION
**Pull request by m-kinkade**
# Critical Changes

# Changes
### Updates to Ensure Data Integrity
We've added functionality to prevent the deletion of Test records with related Test Score records.

These changes are for internal-use only and are disabled at this time. You won't see any difference in functionality.

**Note:** We strongly recommend that EDA Admins don't manage Hierarchy Custom Settings directly and, instead, use the EDA Settings UI, which will be provided in an upcoming release. However, if you do enable any of these settings, note the following:

- If you have disabled the related TST_CannotDelete_TDTM classes for the object, the settings will remain disabled even though you've enabled them in Hierarchy Custom Settings.

- If you enable any of these settings, be sure to disable them before merging records or deleting duplicate records.

# Issues Closed
Closes Issue #1170 

# New Metadata
Checkbox Field on Hierarchy Settings:

- Prevent_Test_Deletion__c

Apex Classes:

- TST_CannotDelete_TDTM

- TST_CannotDelete_TEST

# Deleted Metadata

# Testing Notes
https://salesforce.quip.com/tu19APFVFuvF#OGIACAhkxdd

### CODE REVIEW

File: src/classes/TST_CannotDelete_TDTM.cls [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+        
+        if (TST_CannotDelete_TDTM.enabledPreventTestDeletion == FALSE) {
+            return new DmlWrapper(); 
+        }
+        
+        Map<Id, Test__c> oldmap = new Map<Id, Test__c>((List<Test__c>)oldList);
+        
+        if (triggerAction == TDTM_Runnable.Action.BeforeDelete) {
+            for (Test__c tst : [
```
Comment (xinnanlisf) [inline src/classes/TST_CannotDelete_TDTM.cls:62]: The indentation looks a little weird here.

File: src/classes/TST_CannotDelete_TDTM.cls [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    
+    /*******************************************************************************************************
+     * @description Evaluates whether the Test has any child (Test Score) related records.
+     * @param tst is the current Test record.
+     * @return Boolean.
+     ********************************************************************************************************/
+    @testVisible
+    private Boolean hasChildRecords(Test__c tst) {
+
```
Comment (xinnanlisf) [inline src/classes/TST_CannotDelete_TDTM.cls:86]: Same here. I don't think we need this spacing.

File: src/objects/Hierarchy_Settings__c.object [hunk 3]
```diff
@@ -436,6 +436,15 @@
         <trackTrending>false</trackTrending>
         <type>Checkbox</type>
     </fields>
+    <fields>
+        <fullName>Prevent_Test_Deletion__c</fullName>
+        <defaultValue>false</defaultValue>
```
Comment (xinnanlisf) [inline src/objects/Hierarchy_Settings__c.object:441]: I thought we wanted it to be enabled by default?

**Comment by m-kinkade** [comment]
> The PR looks good. Great work @m-kinkade! There are some minor indentation/spacing issues thou. Let's make sure we are consistent with spacing: https://github.com/SalesforceFoundation/EDA/pull/1165/files
> Let me know if any questions!
> 
> Also, I thought we wanted to enable this kind of setting by default now. Maybe that was me getting confused. I will also give a quick round of testing myself.

Thank you for reviewing @xinnanlisf...I'll make the changes now :)

File: src/objects/Hierarchy_Settings__c.object (hunk 3 above)
Comment (m-kinkade) [reply src/objects/Hierarchy_Settings__c.object:450]: My understanding from the spec doc is it should be disabled for new and existing customers: 
"Ensure that Prevent_Test_Deletion__c is disabled for new and existing customers."

I'm happy to update if it was changed.  Is the spec doc the source of truth?

Comment (xinnanlisf) [reply src/objects/Hierarchy_Settings__c.object:459]: that makes sense.

**Comment by spinkelman** [comment]
This passes QA as long as the branch conflict does not affect functional code of the feature. Once DEV confirms I will merge.

**Comment by spinkelman** [comment]
Passes QA - Merging
Will Close WI

**Pull request merged**